ocaml/typing/solver.ml

```diff
@@ -635,6 +635,167 @@ module Solver_mono (C : Lattices_mono) = struct
       end
     end
 
+  (* Moves every reachable variable from [u] such that
+  [current_level] < [u.level] < [generic_level] to
+  [generic_level + (u.level - current_level)], preserving the exact topology *)
+  let rec generalize_topology
+      : type a. log:_ -> a C.obj -> current_level:int ->
+        generic_level:int -> a var -> unit =
+    fun ~log dst ~current_level ~generic_level u ->
+      if u.level <= current_level || u.level >= generic_level then ()
+      else begin
+        let new_level = generic_level + (u.level - current_level) in
+        set_level ~log u new_level;
+        let do_gen (Amorphvar (v, f)) =
+          let src = C.src dst f in
+          generalize_topology ~log src ~current_level ~generic_level v
+        in
+        List.iter do_gen u.vupper;
+        List.iter do_gen u.vlower
+      end
+
+  (* Tightens the bounds of variables based on children whose bounds are equal
+  If a [v, f] in [u.vlower] has bounds that are equal: [f v.lower] <= [u]
+  If a [v, f] in [u.vupper] has bounds that are equal: [u] <= [f v.upper] *)
```

Comment:
That makes sense. It would also improve subsequent calls to "syntactic" comparison of generic variables (submode_ll, etc.) which also need to know the precise bounds rather than conservative bounds

Reply:
I implemented a new version of `generalize_structure` following our offline discussion. I have a few remaining questions: 

1. In my notes we discussed that if a variable is "fully open" nothing should happen. If I recall correctly, "fully open" means there it does not currently have any constraints. Should its bounds also be equal to [min, max]? 
2. Should the variables that are not generalized be updated to the current level? (including both the new copes, and the "fully open" case)

Note that I used the copying function from PR #3150

Reply:
> Should its bounds also be equal to [min, max]?

That's right.

> Should the variables that are not generalized be updated to the current level?

Yes, they should. Any variable that is reachable from the type at the end should either be generic or less than or equal to the current level.